dlup/_background.pyx

```diff
@@ -0,0 +1,150 @@
+                    row_ptr[x + 2] +
+                    row_ptr[x + 3] +
+                    row_ptr[x + 4] +
+                    row_ptr[x + 5] +
+                    row_ptr[x + 6] +
+                    row_ptr[x + 7]
+                )
+
+            # Handle remaining pixels in the row
+            for x in range(aligned_width, width):
+                row_sum += row_ptr[x]
+
+            sum += row_sum
+    else:
+        # Handle case where width < 8
+        for y in range(height):
+            row_ptr = data + y * stride
+            row_sum = 0
+            for x in range(width):
+                row_sum += row_ptr[x]
+            sum += row_sum
+
+    return sum
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _get_foreground_indices_numpy(
+    int image_width,
+    int image_height,
+    float image_slide_average_mpp,
+    np.ndarray background_mask,
+    np.ndarray[np.float64_t, ndim=2] regions_array,
+    float threshold,
+    np.ndarray[np.int64_t, ndim=1] foreground_indices,
+):
+    cdef:
+        int idx, num_regions
+        float x, y, w, h, mpp
+        int width, height, max_dimension
+        float image_slide_scaling
+        int region_width, region_height
+        float scale_factor
+        int clipped_w, clipped_h
+        int x1, y1, x2, y2
+        unsigned char[:, ::1] background_mask_view
+        const unsigned char* background_mask_ptr
+        const unsigned char* mask_tile_ptr
+        Py_ssize_t mask_tile_stride
+        long long sum_value
+        int error_flag = 0  # 0: No error, 1: MPP zero, 2: Region size zero
+        int foreground_count = 0
+
+    num_regions = regions_array.shape[0]
+    if foreground_indices.shape[0] < num_regions:
+        raise ValueError("foreground_indices array must be at least as long as regions_array")
+
+    background_mask_view = background_mask
+    background_mask_ptr = &background_mask_view[0, 0]
+    mask_tile_stride = background_mask_view.strides[0]
+
+    height, width = background_mask.shape[:2]
+    max_dimension = max_c(width, height)
+
+    for idx in range(num_regions):
+        x, y = regions_array[idx, 0], regions_array[idx, 1]
+        w, h = regions_array[idx, 2], regions_array[idx, 3]
+        mpp = regions_array[idx, 4]
+
+        if mpp == 0:
+            error_flag = 1
+            break
+
+        image_slide_scaling = safe_divide(image_slide_average_mpp, mpp)
+        region_width = <int>(image_slide_scaling * image_width)
+        region_height = <int>(image_slide_scaling * image_height)
+
+        if region_width == 0 or region_height == 0:
+            error_flag = 2
+            break
+
+        scale_factor = safe_divide(max_dimension, <float>max_c(region_width, region_height))
+
+        x1 = min_c(width, <int> floor(x * scale_factor))
+        y1 = min_c(height, <int> floor(y * scale_factor))
+        x2 = min_c(width, <int> ceil((x + w) * scale_factor))
+        y2 = min_c(height, <int> ceil((y + h) * scale_factor))
```

Comment:
Should we make a check here whether x1,y1 are smaller than 0, or x2,y2 > width, height? If we do go out of bounds there is no check right now.